staging/src/k8s.io/apiserver/pkg/server/healthz.go

```diff
@@ -18,28 +18,117 @@ package server
 
 import (
 	"fmt"
-
+	"k8s.io/apimachinery/pkg/util/clock"
 	"k8s.io/apiserver/pkg/server/healthz"
+	"net/http"
+	"time"
 )
 
-// AddHealthzCheck allows you to add a HealthzCheck.
+// AddHealthzCheck adds HealthzCheck(s) to both healthz and readyz. All healthz checks
+// are automatically added to readyz, since we want to avoid the situation where the
+// apiserver is ready but not live.
 func (s *GenericAPIServer) AddHealthzChecks(checks ...healthz.HealthzChecker) error {
 	s.healthzLock.Lock()
 	defer s.healthzLock.Unlock()
 
-	if s.healthzCreated {
-		return fmt.Errorf("unable to add because the healthz endpoint has already been created")
+	if s.healthChecksLocked || s.readinessChecksLocked {
+		return fmt.Errorf("unable to add because the healthz or readyz endpoint has already been created")
 	}
+	s.healthChecks = append(s.healthChecks, checks...)
+	s.readinessChecks = append(s.readinessChecks, checks...)
+	return nil
+}
+
+// AddReadyzChecks allows you to add a HealthzCheck to readyz.
+func (s *GenericAPIServer) AddReadyzChecks(checks ...healthz.HealthzChecker) error {
+	s.healthzLock.Lock()
+	defer s.healthzLock.Unlock()
 
-	s.healthzChecks = append(s.healthzChecks, checks...)
+	if s.readinessChecksLocked {
+		return fmt.Errorf("unable to add because the readyz endpoint has already been created")
+	}
+
+	s.readinessChecks = append(s.readinessChecks, checks...)
 	return nil
 }
 
 // installHealthz creates the healthz endpoint for this server
 func (s *GenericAPIServer) installHealthz() {
 	s.healthzLock.Lock()
 	defer s.healthzLock.Unlock()
-	s.healthzCreated = true
+	s.healthChecksLocked = true
 
-	healthz.InstallHandler(s.Handler.NonGoRestfulMux, s.healthzChecks...)
+	healthz.InstallHandler(s.Handler.NonGoRestfulMux, s.healthChecks...)
 }
+
+// installReadyz creates the readyz endpoint for this server.
+func (s *GenericAPIServer) installReadyz(stopCh <-chan struct{}) {
+	s.AddReadyzChecks(shutdownCheck{stopCh})
+	s.healthzLock.Lock()
+	defer s.healthzLock.Unlock()
+
+	s.readinessChecksLocked = true
+
+	healthz.InstallReadyzHandler(s.Handler.NonGoRestfulMux, s.readinessChecks...)
+}
+
+// shutdownCheck fails if the embedded channel is closed. This is intended to allow for graceful shutdown sequences
+// for the apiserver.
+type shutdownCheck struct {
+	StopCh <-chan struct{}
+}
+
+func (shutdownCheck) Name() string {
+	return "shutdown"
+}
+
+func (c shutdownCheck) Check(req *http.Request) error {
+	select {
+	case <-c.StopCh:
+		return fmt.Errorf("process is shutting down")
+	default:
+	}
+	return nil
+}
+
+
+// AddHealthzCheck adds HealthzCheck(s) to both healthz and readyz. All healthz checks
+// are automatically added to readyz, since we want to avoid the situation where the
+// apiserver is ready but not live.
+func (s *GenericAPIServer) AddDelayedHealthzCheck(check healthz.HealthzChecker) error {
```

Comment:
add a delay parameter and call this with delay=0 from AddHealthzCheck to unify the code. Also move them next to each other to make it easier to maintain.